include/fmt/format.h

```diff
@@ -1822,85 +1843,201 @@ class specs_handler: public specs_setter<typename Context::char_type> {
   FMT_CONSTEXPR void on_dynamic_precision(Id arg_id) {
-    specs_.precision_ref = make_arg_ref(arg_id);
+    context_.check_arg_id(arg_id);
+    specs_.precision_ref = arg_ref_creator_.create(arg_id);
   }
 
   FMT_CONSTEXPR void on_error(const char *message) {
     context_.on_error(message);
   }
 
  private:
-  typedef arg_ref<char_type> arg_ref_type;
+  Specs &specs_;
+  ParseContext &context_;
+  ArgRefCreator &arg_ref_creator_;
+};
 
-  template <typename Id>
-  FMT_CONSTEXPR arg_ref_type make_arg_ref(Id arg_id) {
-    context_.check_arg_id(arg_id);
-    return arg_ref_type(arg_id);
+template <typename ParseContext>
+class dynamic_arg_ref_creator {
+  typedef typename ParseContext::char_type char_type;
+
+ public:
+  typedef arg_ref<char_type, string_value<char_type>> arg_ref_type;
+
+  FMT_CONSTEXPR dynamic_arg_ref_creator(ParseContext &ctx) : context_(ctx) {}
+
+  FMT_CONSTEXPR arg_ref_type create(unsigned id) { return arg_ref_type(id); }
+
+  FMT_CONSTEXPR arg_ref_type create(basic_string_view<char_type> id) {
+    return arg_ref_type(string_value<char_type>{id.data(), id.size()});
   }
 
-  FMT_CONSTEXPR arg_ref_type make_arg_ref(auto_id) {
+  FMT_CONSTEXPR arg_ref_type create(auto_id) {
     return arg_ref_type(context_.next_arg_id());
   }
 
-  dynamic_format_specs<char_type> &specs_;
+ private:
   ParseContext &context_;
 };
 
+struct string_view_metadata {
+  FMT_CONSTEXPR string_view_metadata() : offset_(0u), size_(0u) {}
+  template <typename Char>
+  FMT_CONSTEXPR string_view_metadata(basic_string_view<Char> primary_string,
+                                     basic_string_view<Char> view)
+      : offset_(static_cast<unsigned>(view.data() - primary_string.data())),
+        size_(static_cast<unsigned>(view.size())) {}
+  FMT_CONSTEXPR string_view_metadata(unsigned offset, unsigned size)
+      : offset_(offset), size_(size) {}
+  template <typename S>
+  FMT_CONSTEXPR typename std::enable_if<internal::is_string<S>::value,
+                                        basic_string_view<FMT_CHAR(S)>>::type
+  to_view(S &&str) const {
+    const auto view = to_string_view(str);
+    return basic_string_view<FMT_CHAR(S)>(view.data() + offset_, size_);
+  }
+  unsigned offset_;
+  unsigned size_;
+};
+
+template <typename ParseContext>
+class prepared_arg_ref_creator {
+  typedef typename ParseContext::char_type char_type;
+
+ public:
+  typedef arg_ref<char_type, string_view_metadata> arg_ref_type;
+
+  FMT_CONSTEXPR
+  prepared_arg_ref_creator(ParseContext &ctx,
+                           basic_string_view<char_type> format_str)
+      : context_(ctx), format_(format_str) {}
+
+  FMT_CONSTEXPR arg_ref_type create(basic_string_view<char_type> id) {
+    const auto id_metadata = string_view_metadata(format_, id);
+    return arg_ref_type(id_metadata);
+  }
+
+  FMT_CONSTEXPR arg_ref_type create(unsigned id) { return arg_ref_type(id); }
+
+  FMT_CONSTEXPR arg_ref_type create(auto_id) {
+    return arg_ref_type(context_.next_arg_id());
```

Comment:
As above, `next_arg_id` should be moved up the stack and parse context dependency removed.